dlog/plonk/src/plonk_sponge.rs

```diff
@@ -0,0 +1,45 @@
+use crate::prover::ProofEvaluations;
+use algebra::{
+    Field, PrimeField,
+};
+use oracle::poseidon::{ArithmeticSponge, ArithmeticSpongeParams, Sponge};
+use oracle::sponge::{DefaultFrSponge, ScalarChallenge};
+
+pub trait FrSponge<Fr: Field> {
+    fn new(p: ArithmeticSpongeParams<Fr>) -> Self;
+    fn absorb(&mut self, x: &Fr);
+    fn challenge(&mut self) -> ScalarChallenge<Fr>;
+    fn absorb_evaluations(&mut self, x_hat_beta1: &[Fr], e: &ProofEvaluations<Fr>);
+}
+
+impl<Fr: PrimeField> FrSponge<Fr> for DefaultFrSponge<Fr> {
+    fn new(params: ArithmeticSpongeParams<Fr>) -> DefaultFrSponge<Fr> {
+        DefaultFrSponge {
+            params,
+            sponge: ArithmeticSponge::new(),
+            last_squeezed: vec![],
+        }
+    }
+
+    fn absorb(&mut self, x: &Fr) {
+        self.last_squeezed = vec![];
+        self.sponge.absorb(&self.params, &[*x]);
+    }
+
+    fn challenge(&mut self) -> ScalarChallenge<Fr> {
+        ScalarChallenge(self.squeeze(oracle::sponge::CHALLENGE_LENGTH_IN_LIMBS))
+    }
+
+    fn absorb_evaluations(&mut self, x_hat: &[Fr], e: &ProofEvaluations<Fr>) {
+        self.last_squeezed = vec![];
+        self.sponge.absorb(&self.params, x_hat);
+
+        let points = [
```

Comment:
Guessing all the dlog stuff is TODO? It would be good to move it into another PR so we can review and then merge this one.

Reply:
Agreed. Will create another PR over the same branch for dlog-based Plonk implementation..